beacon-chain/state/stateutil/state_hasher.go

```diff
@@ -0,0 +1,547 @@
+package stateutil
+
+import (
+	"context"
+	"encoding/binary"
+	"sync"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/hash"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/encoding/ssz"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"go.opencensus.io/trace"
+)
+
+var (
+	// Set the map size as equal to that of the latest state field count.
+	leavesCache = make(map[string][][32]byte, params.BeaconConfig().BeaconStateMergeFieldCount)
+	layersCache = make(map[string][][][32]byte, params.BeaconConfig().BeaconStateMergeFieldCount)
+	lock        sync.RWMutex
+)
+
+const cacheSize = 100000
+
+// NocachedHasher references a hasher that will not utilize a cache.
+var NocachedHasher *stateRootHasher
+
+// CachedHasher references a hasher that will utilize a roots cache.
+var CachedHasher *stateRootHasher
+
+func init() {
+	rootsCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: cacheSize, // number of keys to track frequency of (1M).
+		MaxCost:     1 << 22,   // maximum cost of cache (3MB).
+		// 100,000 roots will take up approximately 3 MB in memory.
+		BufferItems: 64, // number of keys per Get buffer.
+	})
+	if err != nil {
+		panic(err)
+	}
+	// Temporarily disable roots cache until cache issues can be resolved.
+	CachedHasher = &stateRootHasher{rootsCache: rootsCache}
+	NocachedHasher = &stateRootHasher{}
+}
+
+// stateRootHasher defines an object through which we can
+// hash the different fields in the state with a few cached layers.
+type stateRootHasher struct {
+	rootsCache *ristretto.Cache
+}
+
+// ComputeFieldRootsWithHasherPhase0 hashes the provided phase 0 state and returns its respective field roots.
+func (h *stateRootHasher) ComputeFieldRootsWithHasherPhase0(ctx context.Context, state *ethpb.BeaconState) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherPhase0")
```

Comment:
```suggestion
	_, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherPhase0")
```

deep source complains